### PR TITLE
Added Scrollbar, removed ScrollbarPainter

### PR DESCRIPTION
**Pull request by HansMuller**
Scrollbars use the ScrollNotification machinery to stay in sync with a Scrollable.

Added ScrollNotificationKind so that one can distinghuish the start, middle, and end of scrolls.

Next step: remove OverscrollPainter, add OverscrollIndicator, and remove ScrollableListListener et al.

### CODE REVIEW

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+  final Key scrollableKey;
+  final Widget child;
+
+  @override
+  _ScrollbarState createState() => new _ScrollbarState();
+}
+
+class _ScrollbarState extends State<Scrollbar> {
+  AnimationController _fade = new AnimationController(duration: _kScrollbarThumbFadeDuration);
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: final

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  double _contentExtent;
+
+  @override
+  void initState() {
+    super.initState();
+    _opacity = new CurvedAnimation(parent: _fade, curve: Curves.ease);
+  }
+
+  void updateState(ScrollableState scrollable) {
+    final ExtentScrollBehavior scrollBehavior = scrollable.scrollBehavior;
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: Should we do an `is` check here?

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      updateState(scrollable);
+    });
+  }
+
+  void onScrollEnded(ScrollableState scrollable) {
+    updateState(scrollable);
+    _fade.reverse();
+  }
+
+  bool handleScrollNotification(ScrollNotification notification) {
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: I usually name all these with `_` to emphasize that they're private.

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 4]
```diff
@@ -332,7 +332,7 @@ abstract class ScrollableState<T extends Scrollable> extends State<T> {
       _scrollOffset = newScrollOffset;
     });
     PageStorage.of(context)?.writeState(context, _scrollOffset);
-    new ScrollNotification(this, _scrollOffset).dispatch(context);
+    new ScrollNotification(this, ScrollNotificationKind.updated).dispatch(context);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/scrollable.dart]: Should we move this into `dispatchOnScroll` for symmetry?  As written, you might get an update before a start.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/material/scrollbar.dart (hunk 2 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/scrollbar.dart]: Yes, you're right.  If it's not an ExtentScrollBehavior the update can be short-circuited.

**Comment by HansMuller** [comment]
There's a tricky case here that I've guarded against.

We dispatch the onScrollEnd() and the corresponding ScrollNotification asynchronously, after the fling animation completes. That means that by the time these things run the ScrollableState may have been disposed. 

I'm guarding against that possibility in dispatchOnScrollEnd():
```
  void dispatchOnScrollEnd() {
    assert(_numberOfInProgressScrolls == 0);
    if (config.onScrollEnd != null)
      config.onScrollEnd(_scrollOffset);
    if (context != null)
      new ScrollNotification(this, ScrollNotificationKind.ended).dispatch(context);
  }
```

Perhaps Notification.dispatch() should short-circuit if its target is null, instead?

**Comment by HansMuller** [comment]
Using `mounted` instead of checking to see if the BuildContext is null.

**Pull request merged**